### PR TITLE
Fix recreate masterkey for hsm

### PR DESCRIPTION
**Pull request by sharidas**
Fix recreate masterkey for hsm

Signed-off-by: Sujith H <sharidasan@owncloud.com>

### CODE REVIEW

**Comment by sharidas** [comment]
## Description
When masterkey is enabled with hsm, the recreate mastekey command was failing. The reason for this was Crypt was pointing to wrong class. Meaning instead of `CryptHSM`, it was pointing to `Crypt`. This caused the keys to be generated in the oC instance and hence things were not working. Once the app is re-enabled, the keymanager requires to be pointed to `CryptHSM` if hsm is enabled. The switch to the `Crypt` is decided here https://github.com/owncloud/encryption/blob/master/lib/AppInfo/Application.php#L136-L149.  This pr tries to address this issue.

## Issue
https://github.com/owncloud/encryption/issues/125

## Test done
- Enable encryption with hsm
- Login as admin user
- Verify the OC_DEFAULT_MODULE as shown below:
```
 sujith@sujith-ownCloud  ~/test/owncloud3   master ●  ls -lt data/files_encryption/OC_DEFAULT_MODULE
total 16
-rw-r--r-- 1 sujith sujith 189 May 27 12:35 master_4d9273b1.privateKey
-rw-r--r-- 1 sujith sujith 451 May 27 12:35 master_4d9273b1.publicKey
-rw-rw-r-- 1 sujith sujith 189 May 27 12:35 pubShare_4d9273b1.privateKey
-rw-rw-r-- 1 sujith sujith 451 May 27 12:35 pubShare_4d9273b1.publicKey
 sujith@sujith-ownCloud  ~/test/owncloud3   master ● 
```
- Execute the recreate master key command:
```
✘ sujith@sujith-ownCloud  ~/test/owncloud3   master ●  ./occ encryption:recreate-master-key -y
Cannot load Xdebug - it was already loaded
Decryption started

    1 [->--------------------------]prepare encryption modules...
 done.


 decrypt files for user admin (1 of 1): /admin/files/welcome.txt 
 [-->-------------------------]

 starting to decrypt files... finished 
 [============================]


all files could be decrypted successfully!
    1 [============================]
Decryption completed

Encryption started

Waiting for creating new masterkey

New masterkey created successfully



Encrypt all files with the Default encryption module
====================================================


Use master key to encrypt all files.


Start to encrypt users files
----------------------------



 all files encrypted 
 [============================]


Encryption completed successfully


Note: All users are required to relogin.

 sujith@sujith-ownCloud  ~/test/owncloud3   master ● 
```
- Verify the keys have been updated
```
sujith@sujith-ownCloud  ~/test/owncloud3   master ●  ls -lt data/files_encryption/OC_DEFAULT_MODULE
total 16
-rw-rw-r-- 1 sujith sujith 189 May 27 12:36 master_51aea79b.privateKey
-rw-rw-r-- 1 sujith sujith 451 May 27 12:36 master_51aea79b.publicKey
-rw-rw-r-- 1 sujith sujith 189 May 27 12:36 pubShare_51aea79b.privateKey
-rw-rw-r-- 1 sujith sujith 451 May 27 12:36 pubShare_51aea79b.publicKey
 sujith@sujith-ownCloud  ~/test/owncloud3   master ● 
```
- Verified that files are accessible
- Re-run the recreate masterkey command
```
 sujith@sujith-ownCloud  ~/test/owncloud3   master ●  ./occ encryption:recreate-master-key -y
Cannot load Xdebug - it was already loaded
Decryption started

    1 [->--------------------------]prepare encryption modules...
 done.


 decrypt files for user admin (1 of 1): /admin/files/welcome.txt 
 [-->-------------------------]

 starting to decrypt files... finished 
 [============================]


all files could be decrypted successfully!
    1 [============================]
Decryption completed

Encryption started

Waiting for creating new masterkey

New masterkey created successfully



Encrypt all files with the Default encryption module
====================================================


Use master key to encrypt all files.


Start to encrypt users files
----------------------------



 all files encrypted 
 [============================]


Encryption completed successfully


Note: All users are required to relogin.

 sujith@sujith-ownCloud  ~/test/owncloud3   master ● 
```
- Verified the file welcome.txt was accesible
- Verified the new keys have been generated:
```
 sujith@sujith-ownCloud  ~/test/owncloud3   master ●  ls -lt data/files_encryption/OC_DEFAULT_MODULE
total 16
-rw-rw-r-- 1 sujith sujith 189 May 27 12:36 master_293b1f5e.privateKey
-rw-rw-r-- 1 sujith sujith 451 May 27 12:36 master_293b1f5e.publicKey
-rw-rw-r-- 1 sujith sujith 189 May 27 12:36 pubShare_293b1f5e.privateKey
-rw-rw-r-- 1 sujith sujith 451 May 27 12:36 pubShare_293b1f5e.publicKey
 sujith@sujith-ownCloud  ~/test/owncloud3   master ● 
```

**Comment by jvillafanez** [comment]
There are several things to fix in the class, indirectly related with the work in this PR. I don't know if we can fix them in the PR or not.
* The commands depends on **16** components. This is bad, specially when only 6-8 are directly used by this class. The rest of the components are being passed through. This has several related consequences:
    * The command creates several components by itself. Those components won't be mocked, so *unit tests* are impossible to do. The tests currently made are more in the direction of integration tests.
    * The more components the command relies on, the more probable is that a bug happens. In addition, even if the command itself is simple (in the sense of "I just call this method is we're done"), having more components raises the complexity unnecessarily
* The command overwrites one of the injected components. While the command can be easy to follow, in general we can expect to use the same dependency across the class. It's quite easy to have surprises when you're injecting an object but then you're using a different object.

I'd need to dig into the code to come up with refactor plan, and I'm not sure how much time it could take. For now, I'd recommend to check the possiblity of using at least one factory class to take care of whatever the command needs to create

File: lib/Command/RecreateMasterKey.php [hunk 1]
```diff
@@ -204,24 +151,19 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 
 				$output->writeln("Waiting for creating new masterkey\n");
 
-				$this->keyManager->setPublicShareKeyIDAndMasterKeyId();
-
-				$output->writeln("New masterkey created successfully\n");
-
-				$this->appConfig->setValue('encryption', 'enabled', 'yes');
-				$this->appConfig->setValue('encryption', 'useMasterKey', '1');
+				if ($this->encDecAllFactory->getEncryptAllObj()->createMasterKey($input, $output) === false) {
```
Comment (jvillafanez) [inline lib/Command/RecreateMasterKey.php]: Just create one `EncryptAll` object. You're creating one here and another one below.

File: lib/Command/RecreateMasterKey.php [hunk 2]
```diff
@@ -204,24 +151,19 @@ protected function execute(InputInterface $input, OutputInterface $output) {
...
+					$output->writeln("<error>Error: masterkeys creation failed</error>");
```
Comment (jvillafanez) [inline lib/Command/RecreateMasterKey.php]: Shouldn't we abort the execution here?

File: lib/Crypto/EncryptAll.php [hunk 3]
```diff
@@ -119,6 +119,33 @@ public function __construct(
...
+	 *
+	 * @param InputInterface $input
+	 * @param OutputInterface $output
+	 * @return bool true when masterkey and sharekey is created else false
+	 */
+	public function createMasterKey(InputInterface $input, OutputInterface $output) {
+		$this->keyManager->setPublicShareKeyIDAndMasterKeyId();
+		$output->writeln("New masterkey created successfully\n");
+
+		$this->config->setAppValue('encryption', 'enabled', 'yes');
```
Comment (jvillafanez) [inline lib/Crypto/EncryptAll.php]: This shouldn't be here

File: lib/Crypto/EncryptAll.php [hunk 4]
```diff
@@ -119,6 +119,33 @@ public function __construct(
...
+	/**
+	 * Call this method only when no master key is created.
+	 *
+	 * @param InputInterface $input
+	 * @param OutputInterface $output
+	 * @return bool true when masterkey and sharekey is created else false
+	 */
+	public function createMasterKey(InputInterface $input, OutputInterface $output) {
+		$this->keyManager->setPublicShareKeyIDAndMasterKeyId();
+		$output->writeln("New masterkey created successfully\n");
```
Comment (jvillafanez) [inline lib/Crypto/EncryptAll.php]: This shouldn't be needed. Both the input and output interfaces aren't needed.
Just return true / false to indicate that the keys have been created and validated, and let the caller write the message.

File: lib/Factory/EncDecAllFactory.php [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+		$this->userSession = $userSession;
+	}
+
+	/**
+	 * Returns DecryptAll object
+	 *
+	 * @return DecryptAll
+	 */
+	public function getDecryptAllObj() {
+		return new DecryptAll($this->encryptionManager, $this->userManager, $this->rootView, $this->logger);
```
Comment (jvillafanez) [inline lib/Factory/EncDecAllFactory.php]: It makes more sense to create a new `View` object in order to ensure the `DecryptAll` object has the root view. No need to inject a view.

File: lib/Command/RecreateMasterKey.php (hunk 1 above)
Comment (sharidas) [reply lib/Command/RecreateMasterKey.php]: Done.

File: lib/Command/RecreateMasterKey.php (hunk 2 above)
Comment (sharidas) [reply lib/Command/RecreateMasterKey.php]: Yes. added `return` from the command.

File: lib/Crypto/EncryptAll.php (hunk 3 above)
Comment (sharidas) [reply lib/Crypto/EncryptAll.php]: Removed.

File: lib/Crypto/EncryptAll.php (hunk 4 above)
Comment (sharidas) [reply lib/Crypto/EncryptAll.php]: Removed. And moved to the caller ( in this context recreate master key command )

File: lib/Factory/EncDecAllFactory.php (hunk 5 above)
Comment (sharidas) [reply lib/Factory/EncDecAllFactory.php]: Injected `View` object both in the `EncryptAll` and the `DecryptAll`.

File: lib/Factory/EncDecAllFactory.php [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+	}
+
+	/**
+	 * Returns EncryptAll object
+	 *
+	 * @return EncryptAll
+	 */
+	public function getEncryptAllObj() {
+		$rootView = new View("/");
+		$keyManager = new KeyManager($this->encStorage, $this->getCryptoEngine(),
```
[the remark this answers is not in the document]
Comment (sharidas) [reply lib/Factory/EncDecAllFactory.php]: Added a valid comment.

File: lib/Crypto/EncryptAll.php [hunk 7]
```diff
@@ -119,6 +119,29 @@ public function __construct(
...
 
+	/**
+	 * Call this method only when no master key is created.
+	 *
+	 * @return bool true when masterkey and sharekey is created else false
+	 */
+	public function createMasterKey() {
+		$this->keyManager->setPublicShareKeyIDAndMasterKeyId();
+
+		$this->config->setAppValue('encryption', 'useMasterKey', '1');
```
[the remark this answers is not in the document]
Comment (sharidas) [reply lib/Crypto/EncryptAll.php]: Brought back to the RecreateMasterKey command.

File: tests/unit/Command/RecreateMasterKeyTest.php [hunk 8]
```diff
@@ -226,14 +200,69 @@ function ($path) {
...
-			[false]
+			[true, false],
+			[true, true],
+			[false, false],
+			[false, true]
 		];
 	}
+
+	public function testRecreateMasterkeyFailed() {
+		$this->recreateMasterKey = $this->getMockBuilder('OCA\Encryption\Command\RecreateMasterKey')
```
[the remark this answers is not in the document]
Comment (sharidas) [reply tests/unit/Command/RecreateMasterKeyTest.php]: Rewrote the tests and used command tester for testing.

File: lib/Command/RecreateMasterKey.php [hunk 9]
```diff
@@ -187,41 +125,35 @@ protected function execute(InputInterface $input, OutputInterface $output) {
...
 				}
 				$output->writeln("\nDecryption completed\n");
 
 				//Reencrypt again
 				$this->appManager->enableApp('encryption');
 				$this->appConfig->setValue('core', 'encryption_enabled', 'yes');
-				$this->appConfig->setValue('encryption', 'enabled', 'yes');
+				$this->appConfig->setValue('encryption', 'useMasterKey', '1');
+
 				$output->writeln("Encryption started\n");
```
Comment (jvillafanez) [inline lib/Command/RecreateMasterKey.php]: This message might be misleading. The encryption hasn't started at that point.

File: tests/unit/Command/RecreateMasterKeyTest.php [hunk 10]
```diff
@@ -134,106 +99,83 @@ public function setUp() {
...
+			->willReturn($decryptAll);
+		$this->encdecAllFactory->method('getEncryptAllObj')
+			->willReturn($encryptAll);
+
+		$this->commandTester->execute([
+			'--yes' => 'y'
+		]);
+
+		$output = $this->commandTester->getDisplay();
+		$expectedOutput = "Decryption completed
```
Comment (jvillafanez) [inline tests/unit/Command/RecreateMasterKeyTest.php]: I'd rather check the exit code of the command instead. You can use different constants to check for different errors.

File: lib/Command/RecreateMasterKey.php (hunk 9 above)
Comment (sharidas) [reply lib/Command/RecreateMasterKey.php]: Moved to line above `encryptAll` is called.

File: tests/unit/Command/RecreateMasterKeyTest.php (hunk 10 above)
Comment (sharidas) [reply tests/unit/Command/RecreateMasterKeyTest.php]: Added return values to the command. I know its better to check with the exit code. Somehow I would like an additional assert with output strings.

**Comment by sharidas** [comment]
Backport PR https://github.com/owncloud/core/pull/35380

**Pull request merged**